Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+
+Every one of these hosts serve different scenario and expose different APIs. The one thing they have in common is that their main purpose is to find the right `hostfxr`, load it and call into it to execute the desired scenario. For the most part all these hosts are basically just wrappers around functionality provided by `hostfxr`.
+
+The proposal is to add a new host library `nethost` which can be used by native apps to easily host managed components and to easily locate `hostfxr` for more advanced scenarios.
+
+At the same time add the ability to pass additional runtime properties when starting the runtime through the hosting entry points (starting app, loading component). This can be used by the native app to:
+* Register startup hook without modifying environment variables (which are inherited by child processes)
+* Introduce new runtime knobs which are only available for native hosts without the need to update the hosting APIs every time.
+
+This new ability would be added to both `hostfxr` and to the `nethost`.
+
+
+## New host binary for component hosting
+Add new library `nethost` which will act as the easy to use host for loading managed components.
+The library would be a dynamically loaded library (`.dll`, `.so`, `.dynlib`). For ease of use there would be a header file for C++ apps as well as `.lib`/`.a` for easy linking.
+Apps using the component hosting functionality would ship this library as part ot the app. So similarly to `apphost`/`comhost`/`ijwhost` the `nethost` will be shipped as part of .NET Core SDK to be included in customer's projects.
+
+This library would expose two APIs
+
+### Load managed component and get a function pointer
+``` C++
+int nethost_load_assembly_method(
+        const char * assembly_path,
+        const char * type_name,
+        const char * method_name,
+        const void * reserved,
+        void ** delegate);
+)
+```
+This API will
+* Locate the assembly using the `assembly_path` and its `.runtimeconfig.json` and determine the frameworks it requires to run. (Note that only framework dependent components will be supported for now).
+* If the process doesn't have CoreCLR loaded (more specifically `hostpolicy` library)
+  * Using the `.runtimeconfig.json` resolve required frameworks (just like if running an application) and load the runtime.
+* Else the CoreCLR is already loaded, in that case validate that required frameworks for the component can be satisfied by the runtime.
+  * If the required frameworks are not already present, fail. No support to load additional frameworks for now.
+* Call into the runtime (`System.Private.CoreLib` specifically)
+  * Create a new isolated `AssemblyLoadContext` (possibly reusing for the same components)
+  * Load the component's assembly into it
+  * Find the requested `type_name` and `method_name`
+  * Return a native callable function pointer to the requested method
+
+The `reserved` argument is currently not used and must be set to `nullptr`.
```

Comment:
It might be nice to mention that `assembly_path`, `type_name`, and `method_name` are expected to be utf8 (I assume).

While I saw the request to use the `reserved` for the parameter name, I do not remember why it is in the arguments.  It would be nice to add the rationale for this extra unused parameter.

Reply:
> are expected to be utf8 (I assume).

So many plus ones. All strings should carry a declared encoding.

Reply:
All strings will be the same way they're already in the hosting APIs - that is `pal::char_t *` which means `WCHAR` on Windows (UTF16) and `char` on Linux/macOS (UTF8). I'll add a note about this into the document - I didn't want to "pollute" the proposed APIs with all the syntax sugar necessary for C/C++.